src/types/memoryview.rs

```diff
@@ -0,0 +1,123 @@
+use std::marker::PhantomData;
+use std::os::raw::c_char;
+
+use crate::{AsPyPointer, PyAny, PyResult, Python};
+use crate::buffer::{Element, PyBuffer};
+use crate::ffi::{self, Py_ssize_t};
+use crate::types::PySequence;
+
+/// Represents a Python `memoryview`.
+#[repr(transparent)]
+pub struct PyMemoryView<T>(PyAny, PhantomData<T>);
+
+pyobject_native_type_core!(PyMemoryView<T>, ffi::PyMemoryView_Type, #checkfunction=ffi::PyMemoryView_Check; T);
+
+impl PyMemoryView<u8> {
+    pub fn new<'p>(py: Python<'p>, memory: &mut [u8], is_writable: bool) -> &'p PyMemoryView<u8> {
+        let ptr = memory.as_mut_ptr() as *mut c_char;
+        let len = memory.len() as Py_ssize_t;
+        let flags = if is_writable {
+            ffi::PyBUF_WRITE
+        } else {
+            ffi::PyBUF_READ
+        };
+
+
+        unsafe { py.from_owned_ptr(ffi::PyMemoryView_FromMemory(ptr, len, flags)) }
+    }
+
+    /// Creates a new Python memoryview from a raw pointer and length.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Safety
+    ///
+    /// This function dereferences the raw pointer `ptr` as the
+    /// leading pointer of a slice of length `len`. [As with
+    /// `std::slice::from_raw_parts`, this is
+    /// unsafe](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety).
```

Comment:
Also needs to have a safety note requiring that the created memory view does not outlive the data in `ptr`. This is _extremely_ hard to use safely!